rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/provider/ServerProviderFactory.java

```diff
@@ -211,7 +211,11 @@ protected void setProviders(boolean custom, boolean busGlobal, Object... provide
                 for (Object featureProvider : cfg.getInstances()) {
                     Map<Class<?>, Integer> contracts = cfg.getContracts(featureProvider.getClass());
                     if (contracts != null && !contracts.isEmpty()) {
-                        allProviders.add(new FilterProviderInfo<Object>(featureProvider,
+                        Class<?> providerCls = ClassHelper.getRealClass(getBus(), featureProvider);
```

Comment:
Turned out we handle `ProviderInfo` differently than `FilterProviderInfo`, this minors fixes will match the resource and service class resolution.